smart_open/webhdfs.py

```diff
@@ -28,28 +29,48 @@
 WEBHDFS_MIN_PART_SIZE = 50 * 1024**2  # minimum part size for HDFS multipart uploads
 
 
-def open(uri, mode, min_part_size=WEBHDFS_MIN_PART_SIZE):
+def open(parsed_uri, mode, min_part_size=WEBHDFS_MIN_PART_SIZE):
     """
     Parameters
     ----------
     min_part_size: int, optional
         For writing only.
 
     """
+    http_uri = convert_to_http_uri(parsed_uri)
     if mode == 'rb':
-        return BufferedInputBase(uri)
+        return BufferedInputBase(http_uri)
     elif mode == 'wb':
-        return BufferedOutputBase(uri, min_part_size=min_part_size)
+        return BufferedOutputBase(http_uri, min_part_size=min_part_size)
     else:
-        raise NotImplementedError('webhdfs support for mode %r not implemented' % mode)
+        raise NotImplementedError("webhdfs support for mode %r not implemented" % mode)
+
+
+def convert_to_http_uri(parsed_uri):
+    netloc = parsed_uri.hostname
+    if parsed_uri.port:
+        netloc += ":{}".format(parsed_uri.port)
+    query = parsed_uri.query
+    if parsed_uri.username:
+        query += (
+            ("&" if query else "") + "user.name=" + urlparse.quote(parsed_uri.username)
+        )
+
+    return urlparse.urlunsplit(
+        ("http", netloc, "/webhdfs/v1" + parsed_uri.path, query, "")
+    )
 
 
 class BufferedInputBase(io.BufferedIOBase):
     def __init__(self, uri):
         self._uri = uri
 
         payload = {"op": "OPEN", "offset": 0}
-        self._response = requests.get("http://" + self._uri, params=payload, stream=True)
+        self._response = requests.get(self._uri, params=payload, stream=True)
+        if self._response.status_code != httplib.OK:
+            raise WebHdfsException(
```

Comment:
You're repeating these lines a lot, how about updating the WebHdfsException class to do this for you? I can think of two ways:

```python
raise WebHdfsException.from_response(self._response)  # Static method
raise WebHdfsException(self._response) # Pass directly to constructor
```

Reply:
Good point